File: app/src/main/java/mounts/com/driver/activity/ui/home/HomeFragment.java
//package mounts.com.driver.activity.ui.home;
//
//import android.Manifest;
//import android.annotation.SuppressLint;
//import android.content.Intent;
//import android.content.res.Resources;
//import android.os.Bundle;
//import android.os.Looper;
//import android.provider.Settings;
//import android.util.Log;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.widget.Switch;
//import android.widget.Toast;
//
//import androidx.annotation.NonNull;
//import androidx.core.app.ActivityCompat;
//import androidx.fragment.app.Fragment;
//import androidx.lifecycle.ViewModelProviders;
//
//import com.google.android.gms.location.FusedLocationProviderClient;
//import com.google.android.gms.location.LocationCallback;
//import com.google.android.gms.location.LocationRequest;
//import com.google.android.gms.location.LocationResult;
//import com.google.android.gms.location.LocationServices;
//import com.google.android.gms.maps.CameraUpdate;
//import com.google.android.gms.maps.CameraUpdateFactory;
//import com.google.android.gms.maps.MapsInitializer;
//import com.google.android.gms.maps.OnMapReadyCallback;
//import com.google.android.gms.maps.SupportMapFragment;
//import com.google.android.gms.maps.model.LatLng;
//
//import com.google.android.gms.maps.GoogleMap;
//import com.google.android.gms.maps.model.MapStyleOptions;
//import com.google.android.gms.maps.model.Marker;
//
//import org.kodein.di.Kodein;
//import org.kodein.di.KodeinAware;
//
//import mounts.com.driver.Helpers.Driver;
//import mounts.com.driver.Helpers.FirebaseHelper;
//import mounts.com.driver.Helpers.GoogleMapHelper;
//import mounts.com.driver.Helpers.UiHelper;
//import mounts.com.driver.Interfaces.IPositiveNegativeListener;
//import mounts.com.driver.R;
//import static android.content.ContentValues.TAG;
//
//public class HomeFragment extends Fragment implements OnMapReadyCallback {
//
//    private HomeViewModel homeViewModel;
//    private static final int MY_PERMISSIONS_REQUEST_ACCESS_FINE_LOCATION = 6161;
//    private FusedLocationProviderClient locationProviderClient;
//    private LocationRequest locationRequest;
//     //TextView textView;
//    private   SupportMapFragment mapFragment;
//    private Marker currentPositionMarker = null;
//
//
//
//    private static LocationCallback locationCallback;
//    public static LatLng currentLocation;
//    private boolean locationFlag = true;
//    public static GoogleMap googleMap;
//    private Switch aSwitch;
//    private boolean driverOnlineFlag = false;
//    private GoogleMapHelper googleMapHelper = new GoogleMapHelper();
//    private UiHelper uiHelper = new UiHelper();
//    private FirebaseHelper firebaseHelper = new FirebaseHelper();
//
//    public View onCreateView(@NonNull LayoutInflater inflater,
//                             ViewGroup container, Bundle savedInstanceState) {
//        homeViewModel =
//                ViewModelProviders.of(this).get(HomeViewModel.class);
//        View root = inflater.inflate(R.layout.fragment_home, container, false);
//
//        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
//         mapFragment = (SupportMapFragment)
//                this.getChildFragmentManager()
//                        .findFragmentById(R.id.map);
//        mapFragment.getMapAsync(this);
//
//
//        createLocationCallback();
//        locationProviderClient = LocationServices.getFusedLocationProviderClient(getActivity());
//        locationRequest = uiHelper.getLocationRequest();
//        if (!uiHelper.isPlayServicesAvailable(getContext())) {
//            Toast.makeText(getContext(), "Play Services did not installed!", Toast.LENGTH_SHORT).show();
//            getActivity().finish();
//        } else  requestLocationUpdate();//if(uiHelper.isHaveLocationPermission(getContext()))
//        firebaseHelper.getDriver();
//
//        return root;
//    }
//    @SuppressLint("MissingPermission")
//    private void requestLocationUpdate() {
//        if (!uiHelper.isHaveLocationPermission(getContext())) {
//
//            ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, MY_PERMISSIONS_REQUEST_ACCESS_FINE_LOCATION);
//            return;
//        }
//        if (uiHelper.isLocationProviderEnabled(getContext())) {
//            uiHelper.showPositiveDialogWithListener(getContext(),
//                    getResources().getString(R.string.need_location),
//                    getResources().getString(R.string.location_content),
//                    new IPositiveNegativeListener() {
//                        @Override
//                        public void onNegative() {
//
//                        }
//
//                        @Override
//                        public void onPositive() {
//                            startActivity(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS));
//                        }
//                    }, "Turn On", false);
//        }
//        locationProviderClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper());
//    }
//    public void createLocationCallback() {
//        locationCallback = new LocationCallback() {
//            @Override
//            public void onLocationResult(LocationResult locationResult) {
//                super.onLocationResult(locationResult);
//                if (locationResult.getLastLocation() == null) return;
//                //LatLn
//               currentLocation = new LatLng(locationResult.getLastLocation().getLatitude(), locationResult.getLastLocation().getLongitude());
//
//                if (locationFlag) {
//                    locationFlag = false;
//                    if (currentLocation != null) {
//                        animateCamera(currentLocation);
//
//                        Log.e("Location ***", currentLocation.latitude + " , " + currentLocation.longitude);
//                         showMarker(currentLocation);
//                    } else {
//                       Log.e("currentLocation", "Null");
//
//                    }
//                }//String.format("%d",new Random().nextInt(50)))
//                firebaseHelper.updateDriver(new Driver(currentLocation.latitude, currentLocation.longitude,"lmJMczn0vZMscn2Asq5vSFpCU2l2") );
//
//            }
//        };
//    }
//    private void animateCamera(LatLng latLng) {
//        CameraUpdate cameraUpdate;
//        try {
//            cameraUpdate = googleMapHelper.buildCameraUpdate(latLng);
//            if (cameraUpdate != null) {
//                googleMap.animateCamera(cameraUpdate, 10, null);
//                Log.e("cameraUpdate", "cameraUpdate not null");
//
//            } else if (cameraUpdate == null) {
//                Log.e("AnimateCamera", "cameraupdat null null");
//                googleMap.animateCamera(cameraUpdate, 11, null);
//            }
//        } catch (Exception e) {
//            Log.e("AnimateCamera", e.getMessage());
//        }
//    }
//
//
//    @Override
//    public void onMapReady(GoogleMap gMap) {
//
////        this.googleMap = gMap;
////        MapsInitializer.initialize(getContext());
////        googleMap.setMyLocationEnabled(true);
////        if(currentLocation != null){
////        LatLng sydney = new LatLng(currentLocation.latitude, currentLocation.longitude);
////            googleMap.addMarker(new MarkerOptions().position(sydney).title("Marker in Sydney"));
////            googleMap.moveCamera(CameraUpdateFactory.newLatLng(sydney));
////        }
////        Log.e("Map", "Error");
//        this.googleMap = gMap;
//        if(uiHelper.isHaveLocationPermission(getContext())) {
//            MapsInitializer.initialize(getContext());
//            googleMap.setMyLocationEnabled(true);
//            try {
//                // Customise the styling of the base map using a JSON object defined
//                // in a raw resource file.
//                boolean success = googleMap.setMapStyle(
//                        MapStyleOptions.loadRawResourceStyle(
//                                getContext(), R.raw.style_json_dark));
//
//                if (!success) {
//                    Log.e(TAG, "Style parsing failed.");
//                }
//            } catch (Resources.NotFoundException e) {
//                Log.e(TAG, "Can't find style. Error: ", e);
//            }
//            // Position the map's camera near Sydney, Australia.
//            if (currentLocation != null)
//                googleMap.moveCamera(CameraUpdateFactory.newLatLng(new LatLng(currentLocation.latitude, currentLocation.longitude)));
//        }
//    }
//    private void showMarker(LatLng latLng){
//        try{
//            if (currentPositionMarker == null)
//                currentPositionMarker = googleMap.addMarker(googleMapHelper.getDriverMarkerOptions(latLng));
//            else {
//                Log.e("LagLan", latLng.toString());
//            }
//        }catch (Exception e){
//            Log.e("LagLan", e.getMessage());
//        }
////    if (currentPositionMarker == null)
////        currentPositionMarker = googleMap.
////                addMarker(googleMapHelper.getUserMakerOptions(latLng));
//    }
//
//    @Override
//    public void onStart() {
//        super.onStart();
//        createLocationCallback();
//        requestLocationUpdate();
//        SupportMapFragment mapFragment = (SupportMapFragment)
//                this.getChildFragmentManager()
//                        .findFragmentById(R.id.map);
//        mapFragment.getMapAsync(new OnMapReadyCallback() {
//            @Override
//            public void onMapReady(GoogleMap googleMap) {
//                //this.googleMap = googleMap;
//                if(uiHelper.isHaveLocationPermission(getContext())) {
//                    MapsInitializer.initialize(getContext());
//                    googleMap.setMyLocationEnabled(true);
//                    try {
//                        // Customise the styling of the base map using a JSON object defined
//                        // in a raw resource file.
//                        boolean success = googleMap.setMapStyle(
//                                MapStyleOptions.loadRawResourceStyle(
//                                        getContext(), R.raw.style_json_dark));
//
//                        if (!success) {
//                            Log.e(TAG, "Style parsing failed.");
//                        }
//                    } catch (Resources.NotFoundException e) {
//                        Log.e(TAG, "Can't find style. Error: ", e);
//                    }
//                    // Position the map's camera near Sydney, Australia.
//                    if (currentLocation != null)
//
//                        googleMap.moveCamera(CameraUpdateFactory.newLatLng(new LatLng(currentLocation.latitude, currentLocation.longitude)));
//                }
//            }
//        });
//
//    }
//
//    @Override
//    public void onResume() {
//        super.onResume();
//        SupportMapFragment mapFragment = (SupportMapFragment)
//                this.getChildFragmentManager()
//                        .findFragmentById(R.id.map);
//        mapFragment.getMapAsync(this);
//
//    }
//}

File: app/src/main/java/mounts/com/driver/activity/ui/home/HomeFragment.kt
package mounts.com.driver.activity.ui.home

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.CountDownTimer
import android.provider.Settings
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapsInitializer
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.bottom_sheet_fragment.*
import mounts.com.driver.Helpers.Driver
import mounts.com.driver.R
import mounts.com.driver.Util.GpsUtils
import mounts.com.driver.Util.OnSwipeTouchListener
import mounts.com.driver.databinding.FragmentHomeBinding
import org.kodein.di.KodeinAware
import org.kodein.di.android.x.kodein
import org.kodein.di.generic.instance


class HomeFragment :Fragment(),OnMapReadyCallback,KodeinAware{
    override val kodein by kodein()
    private val factory : HomeViewModelFactory by instance()
    private var isGPSEnabled = false
    lateinit var viewModel:HomeViewModel
     lateinit var binding :FragmentHomeBinding

    companion object{
        @JvmStatic
        lateinit var googleMap:GoogleMap
        @JvmStatic
        private var showBottomSheet:Int? = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        val binding: FragmentHomeBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false)
        viewModel = ViewModelProviders.of(this, factory).get(HomeViewModel::class.java)
        binding.viewModel = viewModel

        showBottomSheet = activity?.intent?.getIntExtra("showBottomSheet",0)
        Log.e("showBottomSheet",showBottomSheet.toString())

        val mapFragment: SupportMapFragment = this.childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
        binding.lifecycleOwner = this

        GpsUtils(context).turnGPSOn(object : GpsUtils.OnGpsListener {

            override fun gpsStatus(isGPSEnable: Boolean) {
                this@HomeFragment.isGPSEnabled = isGPSEnable
            }
        })
        viewModel.getLocationData().observe(this@HomeFragment, Observer {
          //  viewModel.showMarker(LatLng(it.latitude, it.longitude), googleMap)
        })

   if(viewModel.isSingInWithCustomToken()) {
        viewModel.getRealtimeToken().observe(this, Observer { token ->
            if (token != null) {
                viewModel.signInWithCustomToken(token.token.toString())
                Log.e("RealTime SignIn", "Success")
            } else {
                Log.e("RealTime SignIn", "Fail")
            }
        })
    }

        viewModel.getPayloadData().observe(this@HomeFragment, Observer {
            if(it!=null){
                if(showBottomSheet == 1){
                    showBottomSheet = 0
                    countDown()
                }
                if (it.receiver_lat != null && it.receiver_lng != null){
                    viewModel.showUser(LatLng(it.receiver_lat!!.toDouble(),it.receiver_lng!!.toDouble()),googleMap)
                }

                Log.e("Payload","${it}")

            }
        })

        return binding.root

    }


    override fun onStart() {
        super.onStart()
        invokeLocationAction()
        val mapFragment: SupportMapFragment = this.childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
//        viewModel.getLocationData().observe(viewLifecycleOwner, Observer {
//            viewModel.showMarker(LatLng(it.latitude, it.longitude), googleMap)
//        })


    }



    override fun onResume() {
        super.onResume()
        invokeLocationAction()
        val mapFragment: SupportMapFragment = this.childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
        viewModel.getPayloadData().observe(viewLifecycleOwner, Observer {
            if(it!=null){
               // viewModel.showMarker(LatLng(it.receiver_lat!!.toDouble(),it.receiver_lng!!.toDouble()),googleMap)
                Log.e("Payload","${it}")
            }
        })


//        viewModel.getLocationData().observe(this@HomeFragment, Observer {
//        viewModel.showMarker(LatLng(it.latitude, it.longitude), googleMap)
//        })
    }

    fun countDown(){
        val dialogView: View = layoutInflater.inflate(R.layout.bottom_sheet_fragment, null)
        val dialog = context?.let {
            it1 -> BottomSheetDialog(it1,R.style.BottomSheetDialog)
        }

        dialog?.setCanceledOnTouchOutside(false)
        dialog?.setContentView(dialogView)
        dialog?.show()

        var a =0
        val timer = object : CountDownTimer(15000,1000){
            override fun onTick(millisUntilFinished: Long) {
                a= (millisUntilFinished/1000).toInt()
                dialog?.countDownText?.setText("${a}"+" s")
            }
            override fun onFinish() {
                dialog?.countDownText?.setText(a.toString()+" s")
                //TODO:TO UNCOMMENT THIS LINE
               // dialog?.cancel()
            }
        }.start()

    }
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == GPS_REQUEST) {
                isGPSEnabled = true
                invokeLocationAction()
            }
        }
    }
    private fun invokeLocationAction() {
        when {
            !isGPSEnabled -> Log.e("GPS","NEEDED")// buildAlertMessageNoGps("Please Turn on your GPS Connection")

            isPermissionsGranted() -> startLocationUpdate()

            shouldShowRequestPermissionRationale() ->Log.e("Permession","Needed")//buildAlertMessageNoGps(getString(R.string.permission_request))
            else -> ActivityCompat.requestPermissions(
                    requireActivity(),
                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION),
                    LOCATION_REQUEST
            )
        }
    }

    private fun isPermissionsGranted() =
            ActivityCompat.checkSelfPermission(
                    context!!,
                    Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED &&
                    ActivityCompat.checkSelfPermission(
                            context!!,
                            Manifest.permission.ACCESS_COARSE_LOCATION
                    ) == PackageManager.PERMISSION_GRANTED

    private fun startLocationUpdate() {
        viewModel.getLocationData().observe(viewLifecycleOwner, Observer {
            ///latLong.text =  getString(R.string.latLong, it.longitude, it.latitude)
        //    viewModel.showMarker(LatLng(it.latitude,it.longitude),googleMap)
            Log.e("Current Loaction","${it.latitude } ,${it.longitude}")


        })
    }



    private fun shouldShowRequestPermissionRationale() =
            ActivityCompat.shouldShowRequestPermissionRationale(
                    requireActivity(),
                    Manifest.permission.ACCESS_FINE_LOCATION
            ) && ActivityCompat.shouldShowRequestPermissionRationale(
                    requireActivity(),
                    Manifest.permission.ACCESS_COARSE_LOCATION
            )

    @SuppressLint("MissingPermission")
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            LOCATION_REQUEST -> {
                invokeLocationAction()
            }
        }
    }

    fun buildAlertMessageNoGps(title:String) {
        val builder = AlertDialog.Builder(context)
        builder.setMessage(title)
                .setCancelable(false)
                .setPositiveButton("Yes") { dialog, id -> startActivity(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)) }
                .setNegativeButton("No") { dialog, id -> dialog.cancel() }
        val alert = builder.create()
        alert.show()
    }

    override fun onMapReady(p0: GoogleMap?) {

        googleMap = p0!!
        if(isPermissionsGranted()) {
            MapsInitializer.initialize(context)
            googleMap.setMyLocationEnabled(true)

            try {
                // Customise the styling of the base map using a JSON object defined
                // in a raw resource file.
                val success = googleMap.setMapStyle(
                        MapStyleOptions.loadRawResourceStyle(
                                getContext(), R.raw.style_json_dark));

                if (!success) {
                    Log.e("TAG", "Style parsing failed.");
                }
            } catch (e: Exception) {
                Log.e("TAG", "Can't find style. Error: ", e)
            }
            viewModel.getLocationData().observe(viewLifecycleOwner, Observer {
                viewModel.updateDriver(Driver(it.latitude, it.longitude,FirebaseAuth.getInstance().uid.toString()))
             //   viewModel.showMarker(LatLng(it.latitude, it.longitude), googleMap)
                if (it != null)
                    viewModel.animateCamera(LatLng(it.latitude, it.longitude), googleMap)
            })
        }
    }


}
const val LOCATION_REQUEST = 100
const val GPS_REQUEST = 101